File: backend/app/api/routes/customers.py
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status

from app.api.deps import PrincipalDep, SessionDep, require
from app.core.security import Principal

router = APIRouter(prefix="/customers", tags=["Customers"])


@router.get("", status_code=status.HTTP_501_NOT_IMPLEMENTED)
def get_root(
    session: SessionDep,
    principal: Annotated[Principal, Depends(require("customer:read"))],
) -> None:
    """List customers with derived statistics"""
    raise HTTPException(
        status_code=status.HTTP_501_NOT_IMPLEMENTED,
        detail="Not implemented: the service layer for this route is not wired yet.",
    )


@router.post("", status_code=status.HTTP_501_NOT_IMPLEMENTED)
def create_root(
    session: SessionDep,
    principal: Annotated[Principal, Depends(require("customer:write"))],
) -> None:
    """Create a customer"""
    raise HTTPException(
        status_code=status.HTTP_501_NOT_IMPLEMENTED,
        detail="Not implemented: the service layer for this route is not wired yet.",
    )


@router.get("/{customer_id}", status_code=status.HTTP_501_NOT_IMPLEMENTED)
def get_root_2(
    session: SessionDep,
    principal: Annotated[Principal, Depends(require("customer:read"))],
) -> None:
    """One customer"""
    raise HTTPException(
        status_code=status.HTTP_501_NOT_IMPLEMENTED,
        detail="Not implemented: the service layer for this route is not wired yet.",
    )


@router.patch("/{customer_id}", status_code=status.HTTP_501_NOT_IMPLEMENTED)
def update_root(
    session: SessionDep,
    principal: Annotated[Principal, Depends(require("customer:write"))],
) -> None:
    """Update a customer"""
    raise HTTPException(
        status_code=status.HTTP_501_NOT_IMPLEMENTED,
        detail="Not implemented: the service layer for this route is not wired yet.",
    )


@router.get("/{customer_id}/behaviour", status_code=status.HTTP_501_NOT_IMPLEMENTED)
def get_behaviour(
    session: SessionDep,
    principal: Annotated[Principal, Depends(require("customer:read"))],
) -> None:
    """Days beyond terms, by month"""
    raise HTTPException(
        status_code=status.HTTP_501_NOT_IMPLEMENTED,
        detail="Not implemented: the service layer for this route is not wired yet.",
    )

File: backend/src/routes/customers.js
import { Router } from "express";

import { customersController } from "../controllers/customers.js";
import { authenticate } from "../middleware/authenticate.js";
import { requirePermission } from "../middleware/require.js";

export function customersRouter(sql, config) {
  const controller = customersController(sql);
  const router = Router();
  router.use(authenticate(config.secretKey));

  router.get("/", requirePermission("customer:read"), controller.list);
  router.post("/", requirePermission("customer:write"), controller.create);
  router.get("/:customerId", requirePermission("customer:read"), controller.get);
  router.patch("/:customerId", requirePermission("customer:write"), controller.update);
  router.get(
    "/:customerId/behaviour",
    requirePermission("customer:read"),
    controller.behaviour,
  );
  router.get(
    "/:customerId/events",
    requirePermission("customer:read"),
    controller.events,
  );

  return router;
}
